### PR TITLE
Allow shell expanded .conf values

### PR DESCRIPTION
**Pull request by lrascao**
Allow for shell commmand substitution of .conf values,
only the \`ShellCommand\` form is allowed.

### CODE REVIEW

**Comment by lrascao** [comment]
This one should be merged after #13 and conflicts have been resolved

**Comment by lrascao** [comment]
64c7acadae3d88e86ecfcbc757438cdad4f5c359 has the actual change, 3c8439a551731a580ef52b469a405efc8219180a is needed to bring up the coverage percentage so coveralls green lights this

**Comment by lrascao** [comment]
@lukebakken this is all i have in the pipe for `2.5.0`, proposed merge order is #12 , #13 , pause for rebase and then #14

File: src/cuttlefish_conf.erl
Comment (lukebakken) [inline src/cuttlefish_conf.erl:94]: I can hear the screams of the security review now.

**Review by lukebakken — CHANGES_REQUESTED** [review]
My feeling is that executing arbitrary shell commands will not be appreciated by end-users (cc @michaelklishin)

@lrascao can you give a concrete example of the kinds of problems this change addresses?

**Comment by lrascao** [comment]
> My feeling is that executing arbitrary shell commands will not be appreciated by end-users (cc @michaelklishin)

I hadn't really considered the security aspect of it, i guess mostly due to the fact that a malicious actor that is able to add arbitrary shell commands to the `.conf` file should also be able to do a lot of other things (eg. find the cookie of the node and connect remotely and `os:cmd` becomes available as well, alter the extended start script which is also just shell script), if this is an issue we can probably come up with some alternative to mitigate it

> @lrascao can you give a concrete example of the kinds of problems this change addresses?

Sure, a specific one i'm currently working on is running the Erlang VM on Kubernetes. Without getting into too much detail, k8s offers two features (`configMap` and `downwardAPI`) that allow us to expose the resources that are being requested as files in the Erlang VM container. A typical scenario is tweaking the number of VM schedulers available based on the resource limits imposed (the [adoptingerlang.org site](https://adoptingerlang.org/docs/production/kubernetes/) section on k8s has more detail on this). So here's my vision on this:

- Using the `configMap`/`downwardAPI` features, you expose a file under `etc/conf.d/erlang-vm-total-schedulers` containing a single number which is the desired total number of schedulers. K8s doesn't allow any formatting of this data so the file really only contains a number
- A `erlang-vm-k8s.conf` file is generated in `etc/conf.d` that specifies the `erlang.schedulers.total` parameter as:
```
erlang.schedulers.total = `cat erlang-vm-total-schedulers`
```
- `erlang-vm-k8s.conf` is automatically picked up and used to generate the apropriate `vm.args` that will instruct the VM to only run as many schedulers

If the security restriction is a concern i suppose a viable alternative would be to make use the `$()` [substitution feature](https://github.com/Kyorai/cuttlefish/blob/master/src/cuttlefish_generator.erl#L503) to also support files, it's not as flexible as running arbitrary shell script though, thoughts?

**Comment by lukebakken** [comment]
> If the security restriction is a concern i suppose a viable alternative would be to make use the $() substitution feature to also support files

Yes, let's do that. Please use the same syntax as what `bash` supports which is `$(< path/to/file)`:

```
$ echo 'FILE CONTENTS' > testfile.txt

$ echo $(<testfile.txt)
FILE CONTENTS
```

**Comment by lrascao** [comment]
@lukebakken reworked as agreed

**Comment by lrascao** [comment]
moved back to draft, found this is interacting with the previously existing substitution feature

**Review by lukebakken — COMMENTED** [review]
@lrascao if you have time to review the commits I added to handle some edge-cases, I would appreciate it!

**Comment by lrascao** [comment]
👍  changes look good, let's hope coveralls doesn't come back with more coverage demands

**Review by lukebakken — APPROVED** [review]

**Comment by lrascao** [comment]
@lukebakken that's the end of my pipeline, if there's nothing else that you intent do add could you please cut a `2.5.0` and push it to hex? thanks!

**Comment by lukebakken** [comment]
@lrascao I am blocked by this issue - https://github.com/tsloughter/rebar3_hex/issues/194

**Comment by lukebakken** [comment]
@lrascao https://hex.pm/packages/cuttlefish/2.5.0

**Pull request merged**
